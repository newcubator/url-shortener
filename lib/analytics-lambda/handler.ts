import { S3CreateEvent } from "aws-lambda";
import { S3Client, GetObjectCommand } from "@aws-sdk/client-s3";
import { createInterface } from "node:readline";
import { createGunzip } from "node:zlib";
import { randomUUID } from "node:crypto";
import fetch from "node-fetch";

// TODO add you tracking endpoint
const TRACKING_ENDPOINT = "https://analytics.test.com";

const CLOUD_FRONT_LOG_REGEX = new RegExp(
  "^[^#\n](?<date>(?:(?!\t).)*)\t(?<time>(?:(?!\t).)*)\t(?<edge_location>(?:(?!\t).)*)\t(?<sc_bytes>(?:(?!\t).)*)\t(?<c_ip>(?:(?!\t).)*)\t(?<cs_method>(?:(?!\t).)*)\t(?<cs_host>(?:(?!\t).)*)\t(?<cs_uri_stem>(?:(?!\t).)*)\t(?<sc_status>(?:(?!\t).)*)\t(?<referer>(?:(?!\t).)*)\t(?<user_agent>(?:(?!\t).)*)\t(?<uri_query>(?:(?!\t).)*)\t(?<cookie>(?:(?!\t).)*)\t(?<edge_result_type>(?:(?!\t).)*)\t(?<edge_request_id>(?:(?!\t).)*)\t(?<host_header>(?:(?!\t).)*)\t(?<protocol>(?:(?!\t).)*)\t(?<bytes>(?:(?!\t).)*)\t(?<time_taken>(?:(?!\t).)*)\t(?<forwarded_for>(?:(?!\t).)*)\t(?<ssl_protocol>(?:(?!\t).)*)\t(?<ssl_cipher>(?:(?!\t).)*)\t(?<edge_response_result_type>(?:(?!\t).)*)\t(?<protocol_version>(?:(?!\t).)*)\t(?<fle_status>(?:(?!\t).)*)\t(?<fle_encrypted_fields>(?:(?!\t).)*).*"
);

const client = new S3Client({});

export const handler = async (event: S3CreateEvent) => {
  console.log(`Processing ${event.Records.length} log files`);

  for (const record of event.Records) {
    await processLogFile(record.s3.bucket.name, record.s3.object.key);
  }
};

async function processLogFile(bucket: string, key: string) {
  let response;
  try {
    response = await client.send(
      new GetObjectCommand({
        Bucket: bucket,
        Key: key,
      })
    );
  } catch (error) {
    console.error(`Error while loading log file '${bucket}' '${key}' from s3`);
    console.error(error);
    throw error;
  }

  const lineReader = createInterface({
    input: (response.Body as any).pipe(createGunzip()),
    crlfDelay: Infinity,
  });

  for await (const line of lineReader) {
    await processLogLine(line);
  }
}

async function processLogLine(line: string) {
  const found = line.match(CLOUD_FRONT_LOG_REGEX);

  if (found && found.groups) {
    console.info(`Tracking ${found.groups.cs_uri_stem}`);
    const url = buildTrackingUrl(found.groups);

    try {
      await fetch(url);
    } catch (error) {
      console.error(`Error while calling analytics tool api with '${url}'`);
      console.error(error);
      throw error;
    }
  }
}

/**
 * TODO Build the tracking URL fro your tracking tool. This is an example for matomo.
 *
 * Build a matomo tracking api url.
 *
 * @See https://developer.matomo.org/api-reference/tracking-api
 *
 * For google analytics
 *
 * @See https://developers.google.com/analytics/devguides/collection/protocol/v1
 */
function buildTrackingUrl(log: any) {
  const url = new URL(TRACKING_ENDPOINT);
  const qp = url.searchParams;

  qp.append("idsite", "1");
  qp.append("rec", "1");

  const calledUrl = `${log.protocol}://${log.host_header}${log.cs_uri_stem}`;
  qp.append("url", calledUrl);
  qp.append("rand", randomUUID());
  qp.append("apiv", "1");

  if (log.referer != "-") qp.append("urlref", log.referer);
  const time = log.time.split(":");
  qp.append("h", time[0]);
  qp.append("m", time[1]);
  qp.append("s", time[2]);
  qp.append("ua", log.user_agent);
  qp.append("gt_ms", log.time_taken);

  return url.toString();
}
