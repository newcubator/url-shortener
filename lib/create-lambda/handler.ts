import { PutObjectCommand, S3Client } from "@aws-sdk/client-s3";
import { APIGatewayProxyHandlerV2 } from "aws-lambda";
import { nanoid } from "nanoid";
import { SlackUsecase } from "./slack-usecase";
import { UseCaseInterface } from "./useCaseInterface";

const s3Client = new S3Client({});

// TODO change with your api implementation
const useCase: UseCaseInterface = new SlackUsecase();
export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  if (!event.body) {
    console.error("Command payload missing");
    return {
      statusCode: 400,
    };
  }

  let { url, slug, response_url } = useCase.decodeRequest(event.body);

  if (!url) {
    return {
      statusCode: 200,
      body: JSON.stringify({
        text: "Unfortunately, your command was missing the URL. Please try again with a target URL.",
      }),
    };
  }

  if (!slug) {
    slug = nanoid(6);
  }

  try {
    await s3Client.send(
      new PutObjectCommand({
        Bucket: process.env.BUCKET_NAME,
        Key: slug,
        Body: undefined,
        WebsiteRedirectLocation: url,
      })
    );

    const link = `https://${process.env.BUCKET_NAME}/${slug}`;

    return {
      statusCode: 200,
      body: JSON.stringify({
        text: `Here is your link ${link}`,
      }),
    };
  } catch (error) {
    console.error(error);
    await useCase.handleError(response_url, error);
    throw error;
  }
};
