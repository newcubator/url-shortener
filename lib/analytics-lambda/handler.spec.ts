import { handler } from "./handler";
import { GetObjectCommand, S3Client } from "@aws-sdk/client-s3";
import { createReadStream } from "node:fs";
import { mockClient } from "aws-sdk-client-mock";
import fetch from "node-fetch";
import { randomUUID } from "node:crypto";

jest.mock("node:crypto");
jest.mock("node-fetch");

const s3ClientMock = mockClient(S3Client);
const fetchMock = fetch as any as jest.Mock;
const randomUUIDMock = randomUUID as any as jest.Mock;

randomUUIDMock.mockReturnValue("00000000-0000-0000-0000-000000000000");

it("handles event", async () => {
  s3ClientMock.on(GetObjectCommand).resolves({
    Body: createReadStream("./something.2022-11-16-09.0b3e9009.gz"),
  } as any);

  await handler(SAMPLE_EVENT);

  expect(fetchMock.mock.calls).toMatchSnapshot();
});

it("throws when loading the s3 file fails", async () => {
  s3ClientMock.on(GetObjectCommand).rejects(new Error("some s3 error"));

  await expect(handler(SAMPLE_EVENT)).rejects.toThrowError("some s3 error");
});

it("throws when calling matomo api call fails", async () => {
  s3ClientMock.on(GetObjectCommand).resolves({
    Body: createReadStream("./something.2022-11-16-09.0b3e9009.gz"),
  } as any);
  fetchMock.mockRejectedValueOnce(new Error("some matomo api error"));

  await expect(handler(SAMPLE_EVENT)).rejects.toThrowError("some matomo api error");
});

const SAMPLE_EVENT = {
  Records: [
    {
      eventSource: "aws:s3",
      eventName: "ObjectCreated:Put",
      s3: {
        bucket: {
          name: "test.com-logs",
        },
        object: {
          key: "something.2022-11-16-09.0b3e9009.gz",
        },
      },
    },
  ],
} as any;
