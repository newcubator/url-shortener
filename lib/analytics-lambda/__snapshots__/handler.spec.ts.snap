// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles event 1`] = `
[
  [
    "https://analytics.test.com/?idsite=1&rec=1&url=https%3A%2F%2Ftest.de%2Frrr&rand=00000000-0000-0000-0000-000000000000&apiv=1&h=09&m=24&s=57&ua=Mozilla%2F5.0%2520%28Macintosh%3B%2520Intel%2520Mac%2520OS%2520X%252010_15_7%29%2520AppleWebKit%2F537.36%2520%28KHTML%2C%2520like%2520Gecko%29%2520Chrome%2F107.0.0.0%2520Safari%2F537.36&gt_ms=0.022",
  ],
  [
    "https://analytics.test.com/?idsite=1&rec=1&url=https%3A%2F%2Ftest.de%2Ffavicon.ico&rand=00000000-0000-0000-0000-000000000000&apiv=1&urlref=https%3A%2F%2Ftest.de%2Frrr&h=09&m=24&s=58&ua=Mozilla%2F5.0%2520%28Macintosh%3B%2520Intel%2520Mac%2520OS%2520X%252010_15_7%29%2520AppleWebKit%2F537.36%2520%28KHTML%2C%2520like%2520Gecko%29%2520Chrome%2F107.0.0.0%2520Safari%2F537.36&gt_ms=0.020",
  ],
]
`;
